### ISSUE TITLE
cabal new-repl's ghci can't :load other source files properly

### ISSUE THREAD

**Issue by claudeha** [haskell/cabal#5665]
using

    $ cabal --version
    cabal-install version 2.4.0.0
    compiled using version 2.4.0.1 of the Cabal library 

example session

    $ cabal new-repl 
    Resolving dependencies...
    Build profile: -w ghc-8.4.4 -O1
    In order, the following will be built (use -v for more details):
     - cabal-bug-0.1.0.0 (lib) --enable-profiling (configuration changed)
    Configuring library for cabal-bug-0.1.0.0..
    Preprocessing library for cabal-bug-0.1.0.0..
    GHCi, version 8.4.4: http://www.haskell.org/ghc/  :? for help
    [1 of 1] Compiling Library          ( Library.hs, interpreted )
    Ok, one module loaded.
    *Library> :l Example.hs 
    [1 of 1] Compiling Main             ( Example.hs, interpreted )
    Ok, one module loaded.
    Prelude> :m + Main
    .
    <interactive>:1:1: error:
        attempting to use module ‘cabal-bug-0.1.0.0-inplace:Main’ (Example.hs) which is not loaded
    Prelude> :l Example.hs 
    [1 of 1] Compiling Main             ( Example.hs, interpreted )
    Ok, one module loaded.
    Prelude> :m + Main
    Prelude Main> Main.main
    .
    <interactive>:5:1: error:
        Not in scope: ‘Main.main’
        No module named ‘Main’ is imported.
    Prelude Main> :bro
    main:Main.main :: IO ()
    Prelude Main> :q
    Leaving GHCi.

example cabal project [cabal-bug-0.1.0.0.tar.gz](https://github.com/haskell/cabal/files/2551569/cabal-bug-0.1.0.0.tar.gz):

    $ ls
    cabal-bug.cabal  Example.hs  Executable.hs  Library.hs  LICENSE  Setup.hs

workaround (thanks to dmwit on irc) is to specify an executable for the new-repl:

    $ cabal new-repl exe:cabal-bug
    Build profile: -w ghc-8.4.4 -O1
    In order, the following will be built (use -v for more details):
     - cabal-bug-0.1.0.0 (exe:cabal-bug) --enable-profiling (first run)
    Preprocessing executable 'cabal-bug' for cabal-bug-0.1.0.0..
    GHCi, version 8.4.4: http://www.haskell.org/ghc/  :? for help
    [1 of 1] Compiling Main             ( Executable.hs, interpreted )
    Ok, one module loaded.
    *Main> main
    *Main> :l Example.hs 
    [1 of 1] Compiling Main             ( Example.hs, interpreted )
    Ok, one module loaded.
    *Main> main
    665
    *Main> :q
    Leaving GHCi.

**Comment by ChShersh** [comment]
Probably related issue:

* #5544

**Comment by phadej** [comment]
This isn't related to Current Word Directory. GHCi seems to load the `Example.hs` *fine*, but somehow goes bananas after that. In 5544 the `:load` fails already.

**Comment by chowells79** [comment]
I'm seeing the same behavior (but even worse) when trying to simulate the approach of using the global package registry with ad-hoc files by using v2-repl without a project.

I get the exact same broken behavior with `:load` working then not being able to use the loaded value.  But I can't work around it because I'm working projectless.

It would be really great if I could stop going back to v1-install just to get ad-hoc scripts working.

**Comment by chowells79** [comment]
@phadej Well I'm getting #5544 *also*, but this is the more important bug.

```
~/hask$ cabal v2-repl --build-depends lens --build-depends megaparsec
Resolving dependencies...
Build profile: -w ghc-8.6.1 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (lib) (first run)
Configuring library for fake-package-0..
Preprocessing library for fake-package-0..
Warning: No exposed modules
GHCi, version 8.6.1: http://www.haskell.org/ghc/  :? for help
Prelude> :load AoC0818.hs

<no location info>: error: can't find file: AoC0818.hs
Failed, no modules loaded.
Prelude> :cd ~/hask
Prelude> :load AoC0818.hs
[1 of 1] Compiling Main             ( AoC0818.hs, interpreted )
Ok, one module loaded.
Prelude> :browse Main
data main:Main.Tree = main:Main.Tree [main:Main.Tree] [Int]
main:Main.treeP :: Text.Megaparsec.Parsec () String main:Main.Tree
main:Main.metadata ::
  Control.Lens.Type.Traversal' main:Main.Tree Int
main:Main.children ::
  Control.Lens.Type.Traversal' main:Main.Tree main:Main.Tree
main:Main.leaf ::
  Control.Lens.Type.Fold main:Main.Tree main:Main.Tree
main:Main.metachildren ::
  Control.Lens.Type.Fold main:Main.Tree main:Main.Tree
main:Main.part1 :: main:Main.Tree -> Int
main:Main.part2 :: main:Main.Tree -> Int
main:Main.main :: IO ()
Prelude> import Main

<interactive>:1:1: error:
    attempting to use module ‘fake-package-0-inplace:Main’ (AoC0818.hs) which is not loaded
Prelude>
```

Note that I can use `:browse` to see the contents, but I can't actually use them.  (Code here: https://gist.github.com/chowells79/05ded443ba5e5d9e2eeab2f2882c2eca )

**Comment by hvr** [comment]
@chowells79 @claudeha is this still an issue with cabal 3.0?

**Comment by claudeha** [comment]
Yes, still behaves the same.  I tested with
```
$ cabal --version
cabal-install version 3.0.0.0
compiled using version 3.0.0.0 of the Cabal library 
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.8.1
```
with the base dependency bumped in the cabal file.

**Comment by chowells79** [comment]
@hvr I can't actually test right now.  I've been using your wsl ppr, and it doesn't have those versions yet.

**Comment by chowells79** [comment]
@hvr Ok, got GHC 8.8.1 and cabal 3.0.0 installed - behavior is completely unchanged.  I can `:browse` a module after `:load`ing it, but I can't use its contents.

**Comment by hasufell** [comment]
> https://gist.github.com/chowells79/05ded443ba5e5d9e2eeab2f2882c2eca

This module is missing `module Main where`. If you add it, it works.

**Comment by chowells79** [comment]
It's not missing when it's intentionally left out.  But that narrows the bug's location down significantly - something is going wrong with implicit module name handling.

**Comment by hasufell** [comment]
Yeah, it seems cabal is violating: https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-990005.1

> An abbreviated form of module, consisting only of the module body, is permitted. If this is used, the header is assumed to be ‘module Main(main) where’. If the first lexeme in the abbreviated module is not a {, then the layout rule applies for the top level of the module.

**Comment by treblacy** [comment]
Make up a package name, e.g., foo, so that you can go:

ghci -this-unit-id foo

Then :load on a file that doesn't declare "module" will do this.

cabal-install passes -this-unit-id to ghc/ghci whenever the package or component is a library. And when no package, the temporary fake-package is a library package.

Maybe it is ghci's -this-unit-id bug, or feature. Maybe cabal repl should omit -this-unit-id. I don't know.
